### PR TITLE
Relax TMF destination requirements

### PR DESCRIPTION
**Pull request by librasungirl**

### CODE REVIEW

File: src/core/thread/mle.cpp [hunk 1]
```diff
@@ -3235,6 +3235,11 @@ bool Mle::IsAnycastLocator(const Ip6::Address &aAddress) const
     return memcmp(&mMeshLocal16, &aAddress, kRlocPrefixLength) == 0 && aAddress.mFields.m8[14] == Ip6::Address::kAloc16Mask;
 }
 
+bool Mle::IsMeshLocalAddress(const Ip6::Address &aAddress) const
+{
+    return memcmp(&mMeshLocal16, &aAddress,OT_MESH_LOCAL_PREFIX_SIZE) == 0;
```
Comment (bukepo) [inline src/core/thread/mle.cpp]: nit: style issue here. And how about use Ip6::Address::PrefixMatch()?

File: src/core/thread/thread_netif.cpp [hunk 2]
```diff
@@ -192,21 +192,20 @@ ThreadError ThreadNetif::RouteLookup(const Ip6::Address &source, const Ip6::Addr
...
 
-ThreadError ThreadNetif::TmfFilter(const Message &aMessage, const Ip6::MessageInfo &aMessageInfo)
+ThreadError ThreadNetif::TmfFilter(const Message &aMessage, const Ip6::MessageInfo &aMessageInfo, void *aContext)
 {
     ThreadError error = kThreadError_None;
 
     // A TMF message must comply one of the following rules:
-    // 1. Source address is RLOC or ALOC, and destination address is RLOC, ALOC or realm-local multicast.
-    // 2. Both source and destination addresses are link-local.(for Joiner Entrust)
+    // 1. Source address is RLOC, ALOC, and destination address is MeshLocalAddress(RLOC, ALOC, or ME-ELD) or multicast.
```
Comment (bukepo) [inline src/core/thread/thread_netif.cpp]: `RLOC or ALOC`
`MeshLocalAddress` -> mesh local address

Comment (librasungirl) [inline src/core/thread/thread_netif.cpp]: thanks, updated.

File: src/core/thread/mle.cpp (hunk 1 above)
Comment (librasungirl) [inline src/core/thread/mle.cpp]: thanks, updated.

**Comment by librasungirl** [comment]
Marked as P1 since it would impact all certification scenarios with ML-EID as destination due to the strict rules in #1653

File: src/core/coap/coap_server.hpp [hunk 3]
```diff
@@ -391,8 +392,9 @@ class Server : public CoapBase
      * @param[in]   aInterceptor    A pointer to the interceptor.
      *
      */
-    void SetInterceptor(Interceptor aInterpreter) {
+    void SetInterceptor(Interceptor aInterpreter, void *aContext) {
```
Comment (bukepo) [inline src/core/coap/coap_server.hpp]: missing doc `aContext`.

File: src/core/coap/coap_server.hpp [hunk 4]
```diff
@@ -292,14 +292,15 @@ class Server : public CoapBase
      *
      * @param[in]   aMessage        A reference to the message.
      @ @param[in]   aMessageInfo    A reference to the message info associated with @p aMessage.
+     @ @param[in]   aContext        A pointer to ThreadNetif-specific context.
```
Comment (bukepo) [inline src/core/coap/coap_server.hpp]: I don't think this `aContext` must be `ThreadNetif`

**Comment by thcu-gp** [comment]
The above is a work around, but not a real fix (it still violates section 10.6.2 of the Thread Spec)
After a message M triggered address resolution because the destination address was an ML-EID, M has to be changed in two spots before sending it:
1. Change the MAC80154 address to short address obtained from the address resolution (as it is currently done)
2. Replace the IPv6 address (ML-EID) by the RLOC.
With the above changes the fix in #1653, is valid.

One should also consider not using the ML-EID as source address, because this triggers Address resolution in the receiver before it can reply to the originator's request

**Comment by librasungirl** [comment]
@thcu-gp , The aim of address resolution is to find mesh destination for routing in thread network. And taking the nexthop to the mesh destination as the mac dest address.

The IPv6 destination address in IP header should not be changed during delivery.

I agree that CoAP source address should avoid to take ML-EID as source address, that's what #1732 does.

Thanks.

File: src/core/coap/coap_server.hpp (hunk 3 above)
Comment (librasungirl) [inline src/core/coap/coap_server.hpp]: added, thanks

**Comment by librasungirl** [comment]
@jwhui , not sure if the rules here are still too strict. We don't know if the other stack implementations would also to explicitly specify RLOC as source address for TMF when possible.

**Comment by thcu-gp** [comment]
An application should only use ML-EID addresses (RLOCs are for Thread internal use).
So if a (Thread) CoAP message is sent by an application it will have an ML-EID source address.
Since this is not allowed by Thread (section 10.6.2), there is an impasse. Either
- block those (as done in #1653), but this breaks functionality
- Replace ML-EID by RLOC

@librasungirl I disagree, the aim of the address resolution if to find an RLOC (This is an IP feature), having received an RLOC, allows you to use a short address on the MAC layer (call this a bonus). Said differently, since we cannot route (IP-layer) ML-EIDs, we have to switch to RLOCs. The fact that communication between routers uses 6lowpan meshing is irrelevant.

**Comment by librasungirl** [comment]
@thcu-gp 
We may add interfaces to dispatch the CoAP message to Application process if it shares the Thread Management port(:MM) with TMF in core in the future (as in 10.6.3).
But currently OT core only processes TMF CoAP message which meets the stricter rules in this PR (also meet with the rules in 10.6.2), and simply drops CoAP messages which doesn't meet the rules.

Yes, we cannot route ML-EIDs. But we are not 'switching to RLOCs' via replacing ML-EID with RLOC in IP layer. Instead,  6LowPAN Mesh Header would be added for routing in mesh if the destination is not the direct neighbor. (5.10.1.1). 
Lower layer should not modify the IP header,  but only forward the IP packet (if routable) or drop it (if not routable)

@jwhui ,  just as @thcu-gp points out, if the CoAP message has ML-EID as both source and dest address, it doesn't meet with TMF addressing rules in 10.6.2. Then it should be all right for OT stack to not process such CoAP message.

**Comment by thcu-gp** [comment]
The changes in #1653 are compliant with 10.6.2, the changes in #1733 would allow ML-EIDs

Currently (on our local setup) the ThreadTestHarness test 9.2.4 fails because ML-EIDs are used as the source and destination of a MGMT_GET_ACTIVE.req. The problem lies not in filter, but in the fact that ML-EIDs are used in the first place.

How should an application (only aware of ML-EIDs) use the following OpenThread API call?
```
ThreadError otDatasetSendMgmtActiveGet(uint8_t aTlvTypesLength, uint8_t *aTlvTypes, otIp6Address *aAddress);
```
If it uses an ML-EID for `aAddress`, it would never retrieve the dataset, because the CoAP messages would be dropped.

**Comment by jwhui** [comment]
The changes in #1653 are more strict than what Thread 1.1 Section 10.6.2 specifies.  The changes in #1653 requires the IPv6 source address to be an RLOC or ALOC.  However, Thread 1.1 Section 10.6.2 does not require the IPv6 source address to be RLOC/ALOC if the destination address is an RLOC/ALOC or link-local address.

> A TMF CoAP message sent to the Thread Management UDP port (:MM) MUST comply with at least one of these addressing rules:
> * The IPv6 source address is an RLOC or ALOC.
> * The IPv6 destination address is an RLOC or ALOC. The IPv6 destination address is a Link-Local address.
> * Some of the individual TMF Message definitions will also impose specific, stricter rules for addressing.

Note the "at least one of these" wording above.

For background, the rules in Section 10.6.2 were defined with the goal of allowing an application protocol to utilize the same :MM port.  This implies that any "application" wishing to send/receive TMF messages on the :MM port must be aware of RLOC/ALOC addresses.  See [SPEC-498](https://threadgroup.atlassian.net/browse/SPEC-498) for more discussion around this.

We can argue whether or not Section 10.6.2 should be made more strict, but this needs to happen within the Thread Group.

**Comment by librasungirl** [comment]
Thanks @thcu-gp for pointing out the test issues. 
The issue is because that when the dst is ML-EID, according to the select source rules, either RLOC or ML-EID would be selected as source. And OT core would drop the received CoAP message with ML-EID as source.
I will submit a separate PR to explicitly specify RLOC as the source address for CoAP client messages in core to fix the issue.
Additionally, we may need to confirm with other stack vendors to ensure they don't take ML-EID as both source and dest for certification tests. Let's discuss this within ThreadGroup. Thanks.

**Comment by librasungirl** [comment]
#1743 is submitted to specify RLOC as source for CoAP client messages in core

**Comment by librasungirl** [comment]
Rebase to master to solve the conflict

File: src/core/thread/thread_netif.cpp [hunk 5]
```diff
@@ -192,21 +192,20 @@ ThreadError ThreadNetif::RouteLookup(const Ip6::Address &source, const Ip6::Addr
...
-ThreadError ThreadNetif::TmfFilter(const Message &aMessage, const Ip6::MessageInfo &aMessageInfo)
+ThreadError ThreadNetif::TmfFilter(const Message &aMessage, const Ip6::MessageInfo &aMessageInfo, void *aContext)
 {
     ThreadError error = kThreadError_None;
 
     // A TMF message must comply one of the following rules:
-    // 1. Source address is RLOC or ALOC, and destination address is RLOC, ALOC or realm-local multicast.
-    // 2. Both source and destination addresses are link-local.(for Joiner Entrust)
+    // 1. Source address is RLOC or ALOC, and destination address is mesh local address(that is, RLOC, ALOC, or ML-ELD) or multicast.
+    // 2. Both source and destination addresses are link-local (for Joiner Entrust).
```
Comment (thcu-gp) [inline src/core/thread/thread_netif.cpp]: Isn't this still more strict than section 10.6.2 of the specification?
suggestion:
```
(aMessageInfo.GetPeerAddr().IsRoutingLocator() || aMessageInfo.GetPeerAddr().IsAnycastRoutingLocator()) ||
(aMessageInfo.GetSockAddr().IsRoutingLocator() || aMessageInfo.GetPeerSock().IsAnycastRoutingLocator()) ||
aMessageInfo.GetSockAddr().IsLinkLocal()
```

**Comment by thcu-gp** [comment]
I missed the 'at least one' @jwhui is correct

An application using the OpenThread API, doesn't need to be aware of RLOCs, as long as it can count on the OpenThread stack to use an RLOC as a source addresses in TMF CoAP messages triggered by using the API

File: src/core/coap/coap_server.hpp [hunk 6]
```diff
@@ -292,14 +292,15 @@ class Server : public CoapBase
      *
      * @param[in]   aMessage        A reference to the message.
      @ @param[in]   aMessageInfo    A reference to the message info associated with @p aMessage.
+     * @param[in]   aContext        A pointer to arbitrary context information.
      *
      * @retval  kThreadError_None       Server should continue processing this message, other
      *                                  return values indicates the server should stop processing
      *                                  this message.
-     * @retval  kThreadError_Security   The message does not comply with security rules.
+     * @retval  kThreadError_NotTmf     The CoAP message is not TMF messages.
```
Comment (jwhui) [inline src/core/coap/coap_server.hpp]: nit: "is not TMF messages" -> "is not a TMF message"

Comment (librasungirl) [inline src/core/coap/coap_server.hpp]: Updated. Thanks.

**Comment by librasungirl** [comment]
Roll back to the rules in specification.

**Pull request merged**